src/section/mod.rs

```diff
@@ -8,27 +8,395 @@
+        }
+    }
+
+    pub fn chain(&self) -> &SectionProofChain {
+        &self.chain
+    }
+
+    // Creates the shortest proof chain that includes both the key at `their_knowledge`
+    // (if provided) and the key our current `elders_info` was signed with.
+    pub fn create_proof_chain_for_our_info(
+        &self,
+        their_knowledge: Option<u64>,
+    ) -> SectionProofChain {
+        let first_index = self.elders_info_signing_key_index();
+        let first_index = their_knowledge.unwrap_or(first_index).min(first_index);
+        self.chain.slice(first_index..)
+    }
+
+    pub fn elders_info(&self) -> &EldersInfo {
+        &self.elders_info.value
+    }
+
+    pub fn proven_elders_info(&self) -> &Proven<EldersInfo> {
+        &self.elders_info
+    }
+
+    pub fn is_elder(&self, name: &XorName) -> bool {
+        self.elders_info().elders.contains_key(name)
+    }
+
+    /// Generate a new section info(s) based on the current set of members.
+    /// Returns a set of EldersInfos to vote for.
+    pub fn promote_and_demote_elders(
+        &self,
+        network_params: &NetworkParams,
+        our_name: &XorName,
+    ) -> Vec<EldersInfo> {
+        if let Some((our_info, other_info)) = self.try_split(network_params, our_name) {
+            return vec![our_info, other_info];
+        }
+
+        let expected_elders_map = self.elder_candidates(network_params.elder_size);
+        let expected_elders: BTreeSet<_> = expected_elders_map.keys().collect();
+        let current_elders: BTreeSet<_> = self.elders_info().elders.keys().collect();
+
+        if expected_elders == current_elders {
+            vec![]
+        } else if expected_elders.len() < majority_count(current_elders.len()) {
+            warn!("ignore attempt to reduce the number of elders too much");
+            vec![]
+        } else {
+            let new_info = EldersInfo::new(expected_elders_map, self.elders_info().prefix);
+            vec![new_info]
+        }
+    }
+
+    /// Returns whether the given peer adult or elder.
+    pub fn is_adult_or_elder(&self, name: &XorName) -> bool {
+        self.members.is_adult(name) || self.is_elder(name)
+    }
+
+    // Prefix of our section.
+    pub fn prefix(&self) -> &Prefix {
+        &self.elders_info().prefix
+    }
+
+    pub fn members(&self) -> &SectionPeers {
+        &self.members
+    }
+
+    /// Returns members that are either joined or are left but still elders.
+    pub fn active_members(&self) -> impl Iterator<Item = &Peer> {
+        self.members
+            .all()
+            .filter(move |info| {
+                self.members.is_joined(info.peer.name()) || self.is_elder(info.peer.name())
+            })
+            .map(|info| &info.peer)
+    }
+
+    /// Returns adults from our section.
+    pub fn adults(&self) -> impl Iterator<Item = &Peer> {
+        self.members
+            .adults()
+            .filter(move |peer| !self.is_elder(peer.name()))
```

Comment:
we already called as `adults` , why extra check of `elder` here?

Reply:
Because we want nodes that are adults (age > 4) but not elders. I think are usage of the term `adult` is not consistent. Sometimes it just means a node with age > 4, other times it means non-elder node with age > 4. Will fix it.